src/krb5_plugin/passkey/passkey_utils.c

```diff
@@ -0,0 +1,556 @@
+    if (message == NULL) {
+        return NULL;
+    }
+    memset(message, 0, sizeof(struct sss_passkey_message));
+
+    message->phase = phase;
+    message->state = state == NULL ? NULL : strdup(state);
+    message->data.ptr = data;
+
+    if (state != NULL && message->state == NULL) {
+        sss_passkey_message_free(message);
+        return NULL;
+    }
+
+    return message;
+}
+
+static struct sss_passkey_message *
+sss_passkey_message_from_json(const char *json_str)
+{
+    struct sss_passkey_message *message = NULL;
+    enum sss_passkey_phase phase = 0;
+    const char *state = NULL;
+    void *data = NULL;
+    json_error_t jret;
+    json_t *jdata = NULL;
+    json_t *jroot;
+    int ret;
+
+    jroot = json_loads(json_str, 0, &jret);
+    if (jroot == NULL) {
+        return NULL;
+    }
+
+    ret = json_unpack(jroot, "{s:i, s?:s, s?:o}",
+                     "phase", &phase,
+                     "state", &state,
+                     "data", &jdata);
+    if (ret != 0) {
+        goto done;
+    }
+
+    switch (phase) {
+    case SSS_PASSKEY_PHASE_INIT:
+        data = NULL;
+        break;
+    case SSS_PASSKEY_PHASE_CHALLENGE:
+        data = sss_passkey_challenge_from_json_object(jdata);
+        if (data == NULL) {
+            goto done;
+        }
+        break;
+    case SSS_PASSKEY_PHASE_REPLY:
+        data = sss_passkey_reply_from_json_object(jdata);
+        if (data == NULL) {
+            goto done;
+        }
+        break;
+    default:
+        goto done;
+    }
+
+    message = sss_passkey_message_init(phase, state, data);
+    if (message == NULL && phase == SSS_PASSKEY_PHASE_CHALLENGE) {
+        sss_passkey_challenge_free(data);
+    } else if (message == NULL && phase == SSS_PASSKEY_PHASE_REPLY) {
+        sss_passkey_reply_free(data);
+    }
+
+done:
+    json_decref(jroot);
+    return message;
+}
+
+static char *
+sss_passkey_message_to_json(const struct sss_passkey_message *message)
+{
+    json_t *jroot;
+    json_t *jdata;
+    char *str;
+
+    if (message == NULL) {
+        return NULL;
+    }
+
+    switch (message->phase) {
+    case SSS_PASSKEY_PHASE_INIT:
+        if (message->state != NULL || message->data.ptr != NULL) {
+            return NULL;
+        }
+        jdata = NULL;
+        break;
+    case SSS_PASSKEY_PHASE_CHALLENGE:
+        if (message->state == NULL || message->data.challenge == NULL) {
+            return NULL;
+        }
+
+        jdata = sss_passkey_challenge_to_json_object(message->data.challenge);
+        if (jdata == NULL) {
+            return NULL;
+        }
+        break;
+    case SSS_PASSKEY_PHASE_REPLY:
+        if (message->state == NULL || message->data.reply == NULL) {
+            return NULL;
+        }
+
+        jdata = sss_passkey_reply_to_json_object(message->data.reply);
+        if (jdata == NULL) {
+            return NULL;
+        }
+        break;
+    default:
+        return NULL;
+    }
+
+    jroot = json_pack("{s:i, s:s?, s:o?}",
```

Comment:
It looks like `json_unpack()` cannot handle the JSON NULL value only if read explicitly with `n`. So I would suggest to use `{s:i, s:s*, s:o*}` which would not add the optional attributes at all.